File: frontend/src/types/ocr.ts
export type OcrStatus = 'pending' | 'paused' | 'success' | 'failed' | 'confirmed'
export type OcrMultiImageConsistencyStatus = 'passed' | 'review_required' | 'failed'

export interface OcrMultiImageConsistency {
  status: OcrMultiImageConsistencyStatus
  method: string
  review_required: boolean
  batch_confirm_allowed: boolean
  message: string
  conflicts: Array<Record<string, unknown>>
  llm_judgement?: Record<string, unknown> | null
  llm_error?: string | null
}

export interface OcrMultiImageMeta {
  image_count: number
  merged_from_image_indexes: Record<string, number>
  consistency: OcrMultiImageConsistency
}

/** OCR 提取的结构化药品信息 */
export interface ExtractedDrugData {
  name?: string
  approval_number?: string
  manufacturer?: string
  specification?: string
  batch_number?: string
  production_date?: string  // YYYY-MM-DD 字符串
  expiry_date?: string      // YYYY-MM-DD 字符串
  quantity?: number
  confidence_estimated?: boolean  // true = confidence 是字段完整度代理值
  multi_image?: OcrMultiImageMeta
}

/** OCR 主记录下的单张图片证据 */
export interface OcrRecordImage {
  id: number
  ocr_record_id: number
  image_path: string
  image_index: number
  raw_text?: string
  extracted_data?: ExtractedDrugData
  confidence?: number
  status: OcrStatus
  error_message?: string
  created_at?: string
}

/** OCR 识别记录 */
export interface OcrRecord {
  id: number
  image_path: string
  image_paths?: string[]
  image_count?: number
  images?: OcrRecordImage[]
  raw_text?: string
  extracted_data?: ExtractedDrugData
  confidence?: number
  status: OcrStatus
  drug_id?: number
  batch_id?: number
  error_message?: string
  created_at: string
}

/** 确认入库请求体 */
export interface OcrConfirmRequest {
  drug_id?: number
  drug_name: string
  approval_number?: string
  manufacturer?: string
  specification?: string
  batch_number: string
  production_date?: string  // YYYY-MM-DD
  expiry_date: string       // YYYY-MM-DD（必填）
  quantity: number
  unit: string
}

/** 确认入库响应 */
export interface OcrConfirmResponse {
  ocr_id: number
  drug_id: number
  batch_id: number
  message: string
}

/** 状态对应的中文标签和类型 */
export const OCR_STATUS_MAP: Record<OcrStatus, { label: string; type: string }> = {
  pending:   { label: '识别中',   type: 'info' },
  paused:    { label: '已暂停',   type: 'info' },
  success:   { label: '待确认',   type: 'warning' },
  failed:    { label: '识别失败', type: 'danger' },
  confirmed: { label: '已入库',   type: 'success' },
}

export function isManualReviewRequired(record: OcrRecord): boolean {
  const consistency = record.extracted_data?.multi_image?.consistency
  return consistency?.review_required === true || consistency?.batch_confirm_allowed === false
}

export function isBatchConfirmAllowed(record: OcrRecord): boolean {
  const consistency = record.extracted_data?.multi_image?.consistency
  if (record.status !== 'success') return false
  if (!consistency) return true
  return consistency.batch_confirm_allowed !== false && consistency.review_required !== true
}

File: backend/app/schemas/ocr.py
from datetime import date, datetime
from typing import Any
from pydantic import BaseModel, field_validator

from app.models.ocr_record import OcrStatus


class ExtractedDrugData(BaseModel):
    """OCR 提取的结构化药品信息"""
    name: str | None = None             # 药品名称
    approval_number: str | None = None  # 批准文号
    manufacturer: str | None = None     # 生产厂家
    specification: str | None = None    # 规格
    batch_number: str | None = None     # 批号
    production_date: str | None = None  # 生产日期（原始字符串）
    expiry_date: str | None = None      # 有效期（原始字符串）
    quantity: int | None = None         # 数量


class OcrRecordResponse(BaseModel):
    id: int
    image_path: str
    raw_text: str | None = None
    extracted_data: dict | None = None
    confidence: float | None = None
    status: OcrStatus
    drug_id: int | None = None
    batch_id: int | None = None
    error_message: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class OcrConfirmRequest(BaseModel):
    """用户编辑并确认识别结果，触发入库"""
    # 药品信息：drug_id 有值则关联已有药品；否则创建新药品
    drug_id: int | None = None
    drug_name: str
    approval_number: str | None = None
    manufacturer: str | None = None
    specification: str | None = None

    # 批次信息
    batch_number: str
    production_date: date | None = None
    expiry_date: date
    quantity: int = 0
    unit: str = "盒"

    @field_validator("drug_name")
    @classmethod
    def name_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("药品名称不能为空")
        return v.strip()

    @field_validator("batch_number")
    @classmethod
    def batch_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("批号不能为空")
        return v.strip()


class OcrConfirmResponse(BaseModel):
    """确认入库后返回结果"""
    ocr_id: int
    drug_id: int
    batch_id: int
    message: str = "识别结果已确认入库"


class OcrListQuery(BaseModel):
    """OCR 记录列表查询参数"""
    status: OcrStatus | None = None
    page: int = 1
    page_size: int = 20

    @field_validator("status", mode="before")
    @classmethod
    def normalize_status(cls, v: Any) -> Any:
        """将状态字符串统一转为小写，防止大小写混用导致 422"""
        if isinstance(v, str):
            return v.lower()
        return v
